### PR TITLE
[Narwhal] introduce consensus randomized tests

### PR DESCRIPTION
**Pull request by akichidis**
## Description 

This PR is introducing "randomised" tests for our consensus engine. A main test has been added `bullshark_randomised_tests` where the test orchestration is taking place. Basically the test is running for multiple core properties (ex various DAG rounds, gc_depth, committee size) where for each test iteration it:
* `creates random DAGs` based on some quality characteristics (`FailureModes`)
* `creates multiple execution plans` - different order of feeding the certificates to consensus -  using a variation of [Kahn's algorithm](https://en.wikipedia.org/wiki/Topological_sorting), and executes those via the Bullshark consensus engine
* `compares the output of each execution plan` to ensure they don't diverge - for the same DAG we should expect to have the exact same output for different execution plans

All the random generators used across the above process are seeded by known seeds - and printed on informative output - so we can later easily reproduce a scenario in case of a failure.

The test will run as part of the nightly CI workflow, as it can take quite some time to execute. Also, I have added some finger in the air parameters for the time being, but we can tune them further based on run time etc.

The DAG generation and the execution plan creation is using (seeded) randomness in order to achieve various permutations - which of course is not exhaustive and the different combinations depend to the number of iterations. To reproduce an error someone can use the provided seeds to replay the specific scenario.

### Running the tests manually
To run the tests manually can do via the nextest:
```
 cargo nextest run -p narwhal-consensus bullshark_randomised_tests --run-ignored ignored-only --no-capture
```

If a failure is detected then a log line like the following will be printed:
```
Fork detected in plans for run_id=7, seed=18, rounds=7, committee=4, gc_depth=7, modes=FailureModes { nodes_failure_probability: 0.0, slow_nodes_percentage: 0.33, slow_nodes_failure_probability: 0.7 }'
```

Then the `run_id` can be used to go back to the `bullshark_randomised_tests`, set it to that number and run the test , so can reproduce the result and investigate.

### Running against "fork" bug
As part of https://github.com/MystenLabs/sui/pull/8403 we fixed a fork bug that got introduced in earlier work. When running the randomised tests against the code with the bug managers to successfully detect the fork and report it. To test this I cherry picked the randomised test on top the commit with the issue and ran the tests. The branch can be [found on this branch](https://github.com/MystenLabs/sui/blob/577e722b96925e9161a6d5522ae6024fd937ba3b/narwhal/consensus/src/tests/randomized_tests.rs#L66) .

Running like this:
```
 cargo nextest run -p narwhal-consensus bullshark_randomised_tests --run-ignored ignored-only --no-capture
```

will output:

```
thread 'bullshark::randomized_tests::bullshark_randomised_tests' panicked at 'assertion failed: `(left == right)`
  left: `[dwwml/yqkhsUT8G7: C1(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, eKp5YZF6ApG6lTS9, E0), XwSoL1Tx0XzPywRx: C1(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, RTmx44i8TAq6UJrn, E0), IerZdSvQBNXmie6x: C1(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, OcqdgfvXy0lsqb4p, E0), 0bNKLmBEzcgw+pbR: C2(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, wZtJ2NyeHrAKU7PF, E0), jC8y2AbMI2GMF56d: C1(hsHWobFaIM1WzZYy+mgfv7ABPS0G10Wf2iAWd/xmOla/oarahIEHvNlxCoOnrbteE3O7wsVFwl0HPBZvGCRbdemwudJk1xcPfEc7R9rGmWPP3lkogbw2QK/MVrg2spqG, uvP+CGnDrCSqi4ZP, E0), 1oPpAISn93H9n9ao: C2(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, zVms6gWDMBzVG/Xm, E0), yZ400cpb93I9TnzX: C2(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, TtlXcRyUU2u0+eDQ, E0), 75QIn5xjThJarMUM: C3(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, yUZRIIA32ljxlifT, E0), k2Zqd75XK7Onak9o: C3(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, L1zUn/wKugYLV+Xv, E0), JzxeOnfWcbpAQcts: C3(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, 5ZyjwRmHNMmAOZ/L, E0), fPAjkOz5Tt+mxOZC: C4(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, vJWr4dIDXo0F30Vl, E0), TL8Kydc5sdhSxnP/: C3(hsHWobFaIM1WzZYy+mgfv7ABPS0G10Wf2iAWd/xmOla/oarahIEHvNlxCoOnrbteE3O7wsVFwl0HPBZvGCRbdemwudJk1xcPfEc7R9rGmWPP3lkogbw2QK/MVrg2spqG, fSYvLRNdKrQhcCp+, E0), yMpRyHGQo2cGi/tv: C4(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, ME1XmO723pJhuSqt, E0), ltALm4ODvtotkhyC: C4(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, s80s5UMIrXpXbkXK, E0), nSH6bchZOV93PJho: C4(hsHWobFaIM1WzZYy+mgfv7ABPS0G10Wf2iAWd/xmOla/oarahIEHvNlxCoOnrbteE3O7wsVFwl0HPBZvGCRbdemwudJk1xcPfEc7R9rGmWPP3lkogbw2QK/MVrg2spqG, p4EPG6LVG3S32DMW, E0), f2iy8HuuaxfcspQY: C5(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, cIDSU9Bg3xMrDA4V, E0), fTRna0aA8xNa6N0O: C5(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, sbD8udPdVTQooLuG, E0), HtGRM6uo63WTNzzz: C5(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, rXqqkKkUMUkpmbGJ, E0), rem1TN11AkWXsqaq: C6(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, mXOWLEy6W0f4XGtk, E0)]`,
 right: `[dwwml/yqkhsUT8G7: C1(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, eKp5YZF6ApG6lTS9, E0), XwSoL1Tx0XzPywRx: C1(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, RTmx44i8TAq6UJrn, E0), IerZdSvQBNXmie6x: C1(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, OcqdgfvXy0lsqb4p, E0), 0bNKLmBEzcgw+pbR: C2(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, wZtJ2NyeHrAKU7PF, E0), jC8y2AbMI2GMF56d: C1(hsHWobFaIM1WzZYy+mgfv7ABPS0G10Wf2iAWd/xmOla/oarahIEHvNlxCoOnrbteE3O7wsVFwl0HPBZvGCRbdemwudJk1xcPfEc7R9rGmWPP3lkogbw2QK/MVrg2spqG, uvP+CGnDrCSqi4ZP, E0), 1oPpAISn93H9n9ao: C2(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, zVms6gWDMBzVG/Xm, E0), yZ400cpb93I9TnzX: C2(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, TtlXcRyUU2u0+eDQ, E0), 75QIn5xjThJarMUM: C3(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, yUZRIIA32ljxlifT, E0), k2Zqd75XK7Onak9o: C3(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, L1zUn/wKugYLV+Xv, E0), JzxeOnfWcbpAQcts: C3(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, 5ZyjwRmHNMmAOZ/L, E0), fPAjkOz5Tt+mxOZC: C4(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, vJWr4dIDXo0F30Vl, E0), n5mpACbbClwrs3OK: C2(hsHWobFaIM1WzZYy+mgfv7ABPS0G10Wf2iAWd/xmOla/oarahIEHvNlxCoOnrbteE3O7wsVFwl0HPBZvGCRbdemwudJk1xcPfEc7R9rGmWPP3lkogbw2QK/MVrg2spqG, MeOvYaDWid+gWdAA, E0), TL8Kydc5sdhSxnP/: C3(hsHWobFaIM1WzZYy+mgfv7ABPS0G10Wf2iAWd/xmOla/oarahIEHvNlxCoOnrbteE3O7wsVFwl0HPBZvGCRbdemwudJk1xcPfEc7R9rGmWPP3lkogbw2QK/MVrg2spqG, fSYvLRNdKrQhcCp+, E0), yMpRyHGQo2cGi/tv: C4(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, ME1XmO723pJhuSqt, E0), ltALm4ODvtotkhyC: C4(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, s80s5UMIrXpXbkXK, E0), nSH6bchZOV93PJho: C4(hsHWobFaIM1WzZYy+mgfv7ABPS0G10Wf2iAWd/xmOla/oarahIEHvNlxCoOnrbteE3O7wsVFwl0HPBZvGCRbdemwudJk1xcPfEc7R9rGmWPP3lkogbw2QK/MVrg2spqG, p4EPG6LVG3S32DMW, E0), f2iy8HuuaxfcspQY: C5(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, cIDSU9Bg3xMrDA4V, E0), fTRna0aA8xNa6N0O: C5(qFHWDpkBDWYzUPieD7zvu6vT0eTcUEeRmZ7r9Bd6nVSKZN8/VTXOtBF7I6uTNsZgF0sjkxgE6sTv6rHzYYQXhepALY49pCXzAk1bwd30esNmQxTxKnWOIBl/c5dnnINP, sbD8udPdVTQooLuG, E0), HtGRM6uo63WTNzzz: C5(spJhWejffj1vEBGZgKdW972Zvu/4J0xMhJN9MmHAjGcVqE4VwT13lolJZdTgZHANERCqwhOs1+wZBlG3XfqjwDr8ShhIdxG9x/WJBoihhd7wV6mnw4H37TVyfwc+8hFj, rXqqkKkUMUkpmbGJ, E0), rem1TN11AkWXsqaq: C6(gM7/G81TYckTfJAIEkHv5ituvxNqFP9/ENpMq5+fFkzo+gmIP2GIZNcgbrEKm5e8Ay5FulFJESCqFR70Og9i3kMtvtbzkOA1R4nKyXTWuzAzuvJHo2ezl+5eu/LebwKv, mXOWLEy6W0f4XGtk, E0)]`: Fork detected in plans for run_id=7, seed=18, rounds=7, committee=4, gc_depth=7, modes=FailureModes { nodes_failure_probability: 0.0, slow_nodes_percentage: 0.33, slow_nodes_failure_probability: 0.7 }'
```

where one extra certificate gets committed in one case compared to the other

### Future work
- [ ] Parallelise test runs
- [ ] pretify DAG output and improve the reporting to allow easier debugging
- [ ] add more properties on DAG generator to support more cases
- [ ] currently works when authority per stake = 1, change tests to support variable stakes

## Test Plan 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+    fn hash(&self) -> [u8; crypto::DIGEST_LENGTH] {
+        let mut hasher = crypto::DefaultHashFunction::new();
+        self.certificates.iter().for_each(|c| {
+            hasher.update(c.digest());
+        });
+        hasher.finalize().into()
+    }
+}
+
+#[tokio::test]
```
Comment (mwtian) [inline narwhal/consensus/src/tests/randomized_tests.rs:64]: I have checked with @mlogan that simtests have deterministic random number generations. We can leverage the simtest infrastructure for specifying a seed to produce deterministic results, and starting multiple runs to search for a failing seed. Do you think we should use simtest for the randomized tests here?

Another advantage of using simtest is that it would use the real leader schedule instead of the fake schedule for unit tests.

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+#[ignore]
+async fn bullshark_randomised_tests() {
+    // Configuration regarding the randomized tests. The tests will run for different values
+    // on the below parameters to increase the different cases we can generate.
+
+    // A range of gc_depth to be used
+    const GC_DEPTH: RangeInclusive<Round> = 4..=15;
```
Comment (mwtian) [inline narwhal/consensus/src/tests/randomized_tests.rs:71]: How long does it take to test with more realistic numbers, e.g. GC depth of 50, 100 nodes, 100 ~ 1000 rounds?

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+pub fn make_certificates_with_parameters(
+    seed: u64,
+    committee: &Committee,
+    range: RangeInclusive<Round>,
+    initial_parents: Vec<Certificate>,
+    modes: FailureModes,
+) -> (VecDeque<Certificate>, Vec<Certificate>) {
+    let mut rand = StdRng::seed_from_u64(seed);
+
+    //Pick the slow nodes - ensure we don't have more than 33% of slow nodes
```
Comment (mwtian) [inline narwhal/consensus/src/tests/randomized_tests.rs:215]: nit: space after `//`

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+    // be <=f , otherwise no DAG could be created. The provided number gives the probability of having
+    // failures up to f. Ex for input `failures_probability = 0.2` it means we'll have 20% change of
+    // having failures up to 33% of the nodes.
+    pub nodes_failure_probability: f64,
+
+    // The percentage of slow nodes we want to introduce to our sample. Basically a slow node is one
+    // that might be able to produce certificates, but those are never get referenced by others. That has
+    // as an effect that when they are leaders might also not get enough support - or no support at all.
+    // For example, a value of 0.2 means that we want up to 20% of our nodes to behave as slow nodes.
+    pub slow_nodes_percentage: f64,
```
Comment (mwtian) [inline narwhal/consensus/src/tests/randomized_tests.rs:42]: Is it possible to make every node to have non-zero probability to not be included as a parent? I think it will give more variance to the DAG. But the slow nodes setting may be more realistic. We may be above to simulate commit delay with this.

File: narwhal/consensus/src/tests/randomized_tests.rs (hunk 1 above)
Comment (mystenmark) [reply narwhal/consensus/src/tests/randomized_tests.rs:64]: Yeah, plus that way you can take advantage of other simtest infrastructure, such as the ability to iterate tests, provide seeds on the command line, and the test can become part of the simtest nightly run.

File: narwhal/consensus/src/tests/randomized_tests.rs (hunk 4 above)
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:42]: @mwtian I don't have this set behind a specific probability (although I wanted to - or more specifically chose a distribution that could fit the purpose) but choosing the parents it is also done randomly (uniformly) between a range of `2f + 1  to 3f + 1` . You can see that here https://github.com/MystenLabs/sui/pull/9239/files#diff-8573c7e7cc360f7537b7906716b2e7b99ca41e220054310f9915b47e30bada3dR305

So that would create cases where some parents might not get picked up at all - or picked up by a subset of children certificates (as we currently expect to happen in real life).

File: narwhal/consensus/src/tests/randomized_tests.rs (hunk 2 above)
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:71]: That's a good point , I haven't tested with such large numbers, but to be honest I was thinking that whatever we can achieve with experiments of ~20 rounds we could pretty much achieve with even higher ones. So based on my understanding edge cases could be caught without having the need to run for large round of numbers. I can totally check though.

File: narwhal/consensus/src/tests/randomized_tests.rs (hunk 1 above)
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:64]: @mwtian good suggestion - to be honest I primarily excluded using simtest for the time being as we don't support them for Narwhal + I wanted to better control the seed as all the tests run sequentially (but that might be equally easy with simtets - not aware). @mystenmark let's catch up on what it would take to use simtests for this.

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+#[allow(unused_imports)]
+use tokio::sync::mpsc::channel;
+use types::Round;
+use types::{Certificate, CertificateDigest};
+
+#[derive(Copy, Clone, Debug)]
+pub struct FailureModes {
+    // The probability of having failures per round. The failures should
+    // be <=f , otherwise no DAG could be created. The provided number gives the probability of having
+    // failures up to f. Ex for input `failures_probability = 0.2` it means we'll have 20% change of
```
Comment (LefKok) [inline narwhal/consensus/src/tests/randomized_tests.rs:34]:
```suggestion
    // failures up to f. Ex for input `failures_probability = 0.2` it means we'll have 20% chance of
```

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+    // having failures up to 33% of the nodes.
+    pub nodes_failure_probability: f64,
+
+    // The percentage of slow nodes we want to introduce to our sample. Basically a slow node is one
+    // that might be able to produce certificates, but those are never get referenced by others. That has
+    // as an effect that when they are leaders might also not get enough support - or no support at all.
```
Comment (LefKok) [inline narwhal/consensus/src/tests/randomized_tests.rs:40]:
```suggestion
    // as an effect that when they are leaders they might also not get enough support - or no support at all.
```

File: narwhal/consensus/src/tests/randomized_tests.rs (hunk 2 above)
Comment (LefKok) [reply narwhal/consensus/src/tests/randomized_tests.rs:71]: I would say that most pathogening cases will appear when we do not commit for a long time and then we suddenly commit but different leaders per node. So gc is of little importance since it only works when we do commit

**Review by kchalkias — COMMENTED** [review]
This is a very useful tool for hardening checks, and ideally we need this to be operating. @akichidis do we plan to merge it asap? @mwtian I'd support accepting the deterministic tests as designed by Tasos, without simtests, and then have an upcoming future PR that converts this to the ideal design. My main point is to unblock this as it might help us find last minute edge cases, and we need this asap.

**Comment by akichidis** [comment]
> This is a very useful tool for hardening checks, and ideally we need this to be operating. @akichidis do we plan to merge it asap? @mwtian I'd support accepting the deterministic tests as designed by Tasos, without simtests, and then have an upcoming future PR that converts this to the ideal design. My main point is to unblock this as it might help us find last minute edge cases, and we need this asap.

@kchalkias my plan is actually to merge this without the `simtest` approach and iterate on upcoming PR once we have introduced simtesting for Narwhal.

**Review by kchalkias — APPROVED** [review]
I'm supporting this model for now, even without simtests yet, we can reiterate in the future for extra designs, but ideally we'd like that to be merged asap, so we get any last moment consensus edge case vulnerability signals.
@mystenmark @mwtian

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+        let plan = create_execution_plan(original_certificates.clone(), seed);
+
+        let hash = plan.hash();
+        if !executed_plans.insert(hash) {
+            println!("Skipping plan with seed {}, same executed already", seed);
+            continue;
+        }
+
+        // Now create a new Bullshark engine
+        const NUM_SUB_DAGS_PER_SCHEDULE: u64 = 100;
```
Comment (laura-makdah) [inline narwhal/consensus/src/tests/randomized_tests.rs:422]: It would be nice to have just one of these defined per test package and referenced universally since this value doesn't need to be different for each test

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+use types::Round;
+use types::{Certificate, CertificateDigest};
+
+#[derive(Copy, Clone, Debug)]
+pub struct FailureModes {
+    // The probability of having failures per round. The failures should
+    // be <=f , otherwise no DAG could be created. The provided number gives the probability of having
+    // failures up to f. Ex for input `failures_probability = 0.2` it means we'll have 20% chance of
+    // having failures up to 33% of the nodes.
+    pub nodes_failure_probability: f64,
```
Comment (laura-makdah) [inline narwhal/consensus/src/tests/randomized_tests.rs:36]: it is not clear here what a node failure is?

**Review by arun-koshy — APPROVED** [review]
Excited to see what bugs this catches

File: narwhal/consensus/src/tests/randomized_tests.rs (hunk 8 above)
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:36]: Are you referring to the need for comments? I thought that it would be useful to make things clear to whoever intend to use the tests.

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+        certificates_per_round
+            .entry(c.round())
+            .or_default()
+            .push(c.clone());
+    });
+
+    for round in range {
+        next_parents.clear();
+
+        let mut total_failures = 0;
```
[the remark this answers is not in the document]
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:255]: On this part of the code we are creating the required certificates per round to build up the DAG. For each round we are supposed to (assuming the required probabilities have been set so) introduced some failures (where we basically do not create certificates for some nodes). The `total_failures` variable is tracking the number of failures we have introduced so far in this round. If we have reached `f` , then we stop introducing more as that will lead to create an invalid DAG - or basically we won't be able to advance the DAG to next rounds. The variable is used in lines [266](https://github.com/MystenLabs/sui/pull/9239/files#diff-8573c7e7cc360f7537b7906716b2e7b99ca41e220054310f9915b47e30bada3dR266) and [274](https://github.com/MystenLabs/sui/pull/9239/files#diff-8573c7e7cc360f7537b7906716b2e7b99ca41e220054310f9915b47e30bada3dR274)

**Review by gdanezis — APPROVED** [review]
Very happy to have this facility: two good directions to take this:
(1) lets make sure we support uneven stake to catch issues where.
(2) would love to be able to generate a DAG as a file / other structure and feed it into test cases.

File: narwhal/config/src/committee.rs [hunk 10]
```diff
@@ -177,14 +177,14 @@ impl Committee {
     fn calculate_quorum_threshold(&self) -> NonZeroU64 {
```
[the remark this answers is not in the document]
Comment (akichidis) [reply narwhal/config/src/committee.rs:177]: Good point, I can definitely add those

File: narwhal/consensus/src/tests/randomized_tests.rs (hunk 8 above)
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:36]: Ah got it, let me refine this a bit

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+
+            certificates.push_back(certificate.clone());
+            next_parents.push(certificate);
+        }
+        parents = next_parents.clone();
+    }
+
+    // sanity check - before return, ensure that we have at least 2f+1 certificates per round and
+    // not certificate exists (except the genesis ones) whose parent is missing.
+    certificates_per_round
```
[the remark this answers is not in the document]
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:353]: To be honest I would prefer to keep the check as this method is doing a lot. Now, I could probably change it a little bit and make those checks per round via some more dynamic programming to save cycles, let me re-write this.

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+            for sub_dag in committed_sub_dags {
+                plan_committed_certificates.extend(sub_dag.certificates);
+            }
+        }
+
+        // Compare the results with the previously executed plan results
+        if committed_certificates.is_empty() {
+            committed_certificates = plan_committed_certificates.clone();
+        } else {
+            assert_eq!(
```
[the remark this answers is not in the document]
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:458]: Exactly. As we discussed offline we'll still be covered though as the different execution plans will potentially create sequences that will end up creating partial/subset DAGs. Now, in the end yes everyone will see all the certificates, but in a pathogenic case that wouldn't make a difference - if someone commits a different sub dag we'll still catch this.

**Comment by akichidis** [comment]
> Very happy to have this facility: two good directions to take this:
> (1) lets make sure we support uneven stake to catch issues where.
> (2) would love to be able to generate a DAG as a file / other structure and feed it into test cases.

Thanks @gdanezis for the review. Regarding your points:
(1) Totally agree, I'll do that in a follow up PR as more effort will be needed for this
(2) Even as is someone can use the `generate_randomised_dag` method and create the DAG . That will be in form of sequence of certificates which we can later feed to consensus etc.  Ultimately we could come up with a separate structure that would represent a DAG which could give the ability to generate under such properties - I can look into that separately. Creating as a file we can totally do even as a naive approach (ex serialise on a base64 list or some blob format).

File: narwhal/consensus/src/tests/randomized_tests.rs [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+        .collect();
+
+    // Now shuffle authorities and pick the slow nodes, if should exist
+    ids.shuffle(&mut rand);
+
+    // Step 1 - determine the slow nodes , assuming those should exist
+    let slow_node_ids: Vec<(AuthorityIdentifier, f64)> = {
+        let num_of_slow_nodes =
+            (committee.total_stake() as f64 * modes.slow_nodes_percentage) as Stake;
+        let s = num_of_slow_nodes.min(committee.validity_threshold() - 1);
```
[the remark this answers is not in the document]
Comment (akichidis) [reply narwhal/consensus/src/tests/randomized_tests.rs:230]: Good point , let me do that
